superset-websocket/src/index.ts

```diff
@@ -0,0 +1,465 @@
+// apply config overrides
+Object.assign(opts, config);
+
+// init logger
+const logTransports = [
+  new winston.transports.Console({ handleExceptions: true }),
+];
+if (opts.logToFile && opts.logFilename) {
+  logTransports.push(
+    new winston.transports.File({
+      filename: opts.logFilename,
+      handleExceptions: true,
+    }),
+  );
+}
+const logger = winston.createLogger({
+  level: opts.logLevel,
+  transports: logTransports,
+});
+
+// enforce JWT secret length
+if (startServer && opts.jwtSecret.length < 32)
+  throw new Error('Please provide a JWT secret at least 32 bytes long');
+
+export const redisUrlFromConfig = (redisConfig: RedisConfig): string => {
+  let url = redisConfig.ssl ? 'rediss://' : 'redis://';
+  if (redisConfig.password) url += `:${redisConfig.password}@`;
+  url += `${redisConfig.host}:${redisConfig.port}/${redisConfig.db}`;
+  return url;
+};
+
+// initialize servers
+const redis = new Redis(redisUrlFromConfig(opts.redis));
+const httpServer = http.createServer();
+export const wss = new WebSocket.Server({
+  noServer: true,
+  clientTracking: false,
+});
+
+const SOCKET_ACTIVE_STATES = [WebSocket.OPEN, WebSocket.CONNECTING];
+const GLOBAL_EVENT_STREAM_NAME = `${opts.streamPrefix}full`;
+const DEFAULT_STREAM_LAST_ID = '$';
+
+// initialize internal registries
+export let channels: Record<string, ChannelValue> = {};
+export let sockets: Record<string, SocketInstance> = {};
+let lastFirehoseId: string = DEFAULT_STREAM_LAST_ID;
+
+export const setLastFirehoseId = (id: string): void => {
+  lastFirehoseId = id;
+};
+
+/**
+ * Adds the passed channel and socket instance to the internal registries.
+ */
+export const trackClient = (
+  channel: string,
+  socketInstance: SocketInstance,
+): string => {
+  const socketId = uuidv4();
+  sockets[socketId] = socketInstance;
+
+  if (channel in channels) {
+    channels[channel].sockets.push(socketId);
+  } else {
+    channels[channel] = { sockets: [socketId] };
+  }
+
+  return socketId;
+};
+
+/**
+ * Sends a single async event payload to a single channel.
+ * A channel may have multiple connected sockets, this emits
+ * the event to all connected sockets within a channel.
+ */
+export const sendToChannel = (channel: string, value: EventValue): void => {
+  const strData = JSON.stringify(value);
+  if (!channels[channel]) {
+    logger.debug(`channel ${channel} is unknown, skipping`);
+    return;
+  }
+  channels[channel].sockets.forEach(socketId => {
+    const socketInstance: SocketInstance = sockets[socketId];
+    if (!socketInstance) return cleanChannel(channel);
+    try {
+      socketInstance.ws.send(strData);
+    } catch (err) {
+      logger.debug(`Error sending to socket: ${err}`);
+      // check that the connection is still active
+      cleanChannel(channel);
+    }
+  });
+};
+
+/**
+ * Reads a range of events from a channel-specific Redis event stream.
+ * Invoked in the client re-connection flow.
+ */
+export const fetchRangeFromStream = async ({
```

Comment:
Should this return the reply in a promise instead of using a listener?

Reply:
Yeah, good call. I'll look at refactoring.

Reply:
Thinking about this more, there's an argument to be made for keeping this consistent with the interface for `subscribeToGlobalStream`.